### DISCUSSION TITLE
RavenDb Concurrent Subscription not working

### DISCUSSION THREAD

**Discussion by Janbaz-khan** [ravendb/ravendb#15508] [Q&A]
Hello everyone,
I am new to Raven db I have created a simple POC proving  _Concurrent Subscription in Raven Db_ 
but seems like it is not working for me. below is the code I have written.
Note: I have created a subscription in raven studio with same name.
`
```
DocumentStoreHolder documentStoreHolder = new DocumentStoreHolder();
var store = documentStoreHolder.Store;

// Define concurrent subscription workers
var subscriptionWorker1 = store.Subscriptions.GetSubscriptionWorker<LineItem2>(
    // Set the worker to connect to the "All Orders" subscription task
    new SubscriptionWorkerOptions("All Orders")
    {
                        // Set Concurrent strategy
                        Strategy = SubscriptionOpeningStrategy.Concurrent,
        MaxDocsPerBatch = 20
    });

var subscriptionWorker2 = store.Subscriptions.GetSubscriptionWorker<LineItem2>(
    new SubscriptionWorkerOptions("All Orders")
    {
        Strategy = SubscriptionOpeningStrategy.Concurrent,
        MaxDocsPerBatch = 20
    });
// Start the concurrent worker. Workers will connect concurrently to the "All Orders" subscription task.
var subscriptionRuntimeTask1 = subscriptionWorker1.Run(batch =>
{
                    // process batch
                    foreach (var item in batch.Items)
    {
        Console.WriteLine("Worker 1");
        Console.WriteLine(item.Id);
    }
});

var subscriptionRuntimeTask2 = subscriptionWorker2.Run(batch =>
{
                    // process batch
                    foreach (var item in batch.Items)
    {
        Console.WriteLine("Worker 2");
        Console.WriteLine(item.Id);
    }
});
```

**Comment by ayende** [comment]
What do you mean, not working? I assume that you are seeing only one subscription active here?

Add a `Thread.Sleep(5000)` in the subscription code.
What is likely happening is that the first subscription is fast enough that it is getting all the attention.

**Comment by Janbaz-khan** [comment]
I am seeing "not active" in raven studio. I tried couple of ways to make it
work but it's not working.

**Reply by ayende** [reply]
Are you *waiting* for the subscriptions? They are running in the background.
IF you are not stopping the process from exiting, it won't be around.

**Comment by Janbaz-khan** [comment]
Let me explain my situation.
When I execute the above code. And insert 29 documents in "lineitems"
collection at the same time(using a console app for each loop) ,
Then I should get those 29 documents in batch (in above code) but its not
hitting.

It is working perfectly in "openiffree". But I am planning to execute
concurrent workers for million of documents.

**Reply by ayende** [reply]
Please post your entire code (against live-test.ravendb.net).
Including entity definition and the subscription creation.

**Reply by Janbaz-khan** [reply]
Hi @ayende  I just checked with  a live-test server, and it is working perfectly.

**Reply by Janbaz-khan** [reply]
But unfortunately not working locally, even I test it on my colleague system as well.

**Reply by ayende** [reply]
What version are you running?
Can you post the full code that runs against `localhost:8080` ?

**Reply by Janbaz-khan** [reply]
I am using 5.3 but now used 5.4 as well. still not working.

//*********************************************subscription worker******************************************

`DocumentStoreHolder documentStoreHolder = new DocumentStoreHolder();
var store = documentStoreHolder.Store;

// Define concurrent subscription workers
var subscriptionWorker1 = store.Subscriptions.GetSubscriptionWorker<LineItem2>(
    // Set the worker to connect to the "All Orders" subscription task
    new SubscriptionWorkerOptions("All Orders")
    {
                        // Set Concurrent strategy
                        Strategy = SubscriptionOpeningStrategy.Concurrent,
        MaxDocsPerBatch = 20
    });

var subscriptionWorker2 = store.Subscriptions.GetSubscriptionWorker<LineItem2>(
    new SubscriptionWorkerOptions("All Orders")
    {
        Strategy = SubscriptionOpeningStrategy.Concurrent,
        MaxDocsPerBatch = 20
    });
// Start the concurrent worker. Workers will connect concurrently to the "All Orders" subscription task.
var subscriptionRuntimeTask1 = subscriptionWorker1.Run(batch =>
{
                    // process batch
                    foreach (var item in batch.Items)
    {
        Console.WriteLine("Worker 1");
        Console.WriteLine(item.Id);
    }
});

var subscriptionRuntimeTask2 = subscriptionWorker2.Run(batch =>
{
                    // process batch
                    foreach (var item in batch.Items)
    {
        Console.WriteLine("Worker 2");
        Console.WriteLine(item.Id);
    }
});`

**Reply by ayende** [reply]
Please post *full* code, including the code that you use to create the subscription, including the code to add the documents.
Basically, I want to be able to do `dotnet run Program.cs` and see the behavior

**Reply by Janbaz-khan** [reply]
**Here is the full code including everything**

using DataSubscriptionPOC.Models;
using Raven.Client.Documents;
using Raven.Client.Documents.Subscriptions;
using Raven.Client.ServerWide;
using Raven.Client.ServerWide.Operations;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace DataSubscriptionPOC
{
    public class Program
    {
        static void Main(string[] args)
        {
            //insetin bulk
           // InsertInBulk();
           // 


            DocumentStoreHolder documentStoreHolder = new DocumentStoreHolder();
            var store = documentStoreHolder.Store;
            Console.WriteLine("Starting...");
            Console.WriteLine(store.Conventions.UseOptimisticConcurrency);
            //Creating subscription worker

            //SubscriptionCreationOptions<LineItem2> options = new SubscriptionCreationOptions<LineItem2>();
            //DatabaseTopology topology = store.Maintenance.Server.Send(new GetDatabaseRecordOperation("PocDb")).Topology;
            //options.Name = "LineitemProcessingSubscription";
            //options.MentorNode = topology.Members.First();
            //store.Subscriptions.Create<LineItem2>(options, database: "PocDb");


            //worker 1
            var subscriptionWorker1 = store.Subscriptions.GetSubscriptionWorker<LineItem2>(new SubscriptionWorkerOptions("LineitemProcessingSubscription")
            {
                // Set Concurrent strategy
                MaxDocsPerBatch = 20,
                Strategy = SubscriptionOpeningStrategy.Concurrent

            });

            //worker 2
            var subscriptionWorker2 = store.Subscriptions.GetSubscriptionWorker<LineItem2>(new SubscriptionWorkerOptions("LineitemProcessingSubscription")
            {
                // Set Concurrent strategy
                MaxDocsPerBatch = 20,
                Strategy = SubscriptionOpeningStrategy.Concurrent,
            });

            int counter = 0;
            Console.WriteLine("Worker created with id: " + subscriptionWorker1.WorkerId);
            Console.WriteLine("Worker created with id: " + subscriptionWorker2.WorkerId);
            // Start the concurrent worker. Workers will connect concurrently to the "LineitemProcessingSumbscription" subscription task.
            var subscriptionRuntimeTask1 = subscriptionWorker1.Run(batch =>
            {
                // process batch
                foreach (var item in batch.Items)
                {
                    Console.WriteLine("******************Worker 1****************");
                    Console.WriteLine(item.Id);
                    Console.WriteLine("Exception: " + item.ExceptionMessage);
                    Console.WriteLine("counter: " + counter++);
                }
            });

            // Start the concurrent worker. Workers will connect concurrently to the "LineitemProcessingSumbscription" subscription task.
            var subscriptionRuntimeTask2 = subscriptionWorker2.Run(batch =>
            {
                // process batch
                foreach (var item in batch.Items)
                {
                    Console.WriteLine("******************Worker 2****************");
                    Console.WriteLine(item.Id);
                    Console.WriteLine("Exception: " + item.ExceptionMessage);
                    Console.WriteLine("counter: " + counter++);
                }
            });


            subscriptionWorker1.OnUnexpectedSubscriptionError += x =>
            {
                Console.WriteLine(x.Message);
            };
            subscriptionWorker2.OnUnexpectedSubscriptionError += x =>
           {
               Console.WriteLine(x.Message);
           };

            subscriptionWorker1.OnSubscriptionConnectionRetry += x =>
            {
                Console.WriteLine(x);
            };
            subscriptionWorker2.OnSubscriptionConnectionRetry += x =>
            {
                Console.WriteLine(x);
            };
            Console.ReadLine();
        }
        public static void InsertInBulk()
        {
            DocumentStoreHolder documentStoreHolder = new DocumentStoreHolder();
            var session = documentStoreHolder.Store.OpenSession();
            var random = new Random();
            for (int i = 0; i < 100; i++)//i<10000
            {
                LineItem2 lineItem2 = new LineItem2();
                lineItem2.Name = "name" + random.Next(0, 9999);
                lineItem2.Desctiption = "description" + random.Next(0, 9999);
                lineItem2.Quantity = random.Next(0, 50);
                lineItem2.SerialNumber = random.Next(0, 1000).ToString();
                session.Store(lineItem2);
                if (i != 0 && i % 30 == 0) //savechanges after each 30 records
                {
                    session.SaveChanges();
                    Console.WriteLine(i+". 30 records have been added");
                }
            }
        }
    }
    public class DocumentStoreHolder
    {
        public IDocumentStore Store
        {
            get { return LazyDocStore.Value; }
        }

        private static readonly Lazy<IDocumentStore> LazyDocStore = new Lazy<IDocumentStore>(() =>
        {
            string[] urls = { "http://localhost:8091/" };
            //string[] urls = { "http://live-test.ravendb.net/" };
            var docStore = new DocumentStore();
            docStore.Urls = urls;
            docStore.Database = "PocDb";
            docStore.Conventions.UseOptimisticConcurrency = true;
            docStore.Initialize();
            return docStore;
        });


    }
    public class LineItem2
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string SerialNumber { get; set; }
        public int Quantity { get; set; }
        public string Desctiption { get; set; }
    }
}

**Reply by ayende** [reply]
You aren't `await` ing on the subscription tasks.
The events will be raised only if you are successfully connected, you need to await on the task to see if the connection failed for some readon.

I *assume* that you are using RavenDB 5.4 on the client and an older version on the server, and it throws or something like that.

Here is what I get on my end after running this. 

![image](https://user-images.githubusercontent.com/116915/209784796-1990cc21-43ab-4386-be52-fa004170ecd4.png)

**Comment by Janbaz-khan** [comment]
I will let you know once I post it there.
Thank you so much.
Really appreciate it.
:)

**Comment by Janbaz-khan** [comment]
Thank you for that, Here is what I am seeing.
![exception](https://user-images.githubusercontent.com/37152454/209792053-9c7dbc36-d15d-4d98-9815-026f7245ee0b.PNG)

**Reply by ayende** [reply]
Well, the error explains that pretty well.
You aren't running with concurrent subscription active in your license.
I assume that you are running community?

**Reply by Janbaz-khan** [reply]
correct
